packages/api-plugin-fulfillment-method-shipping-dynamic-rate/src/getFulfillmentMethodsWithQuotesShippingDynamicRate.js

```diff
@@ -0,0 +1,70 @@
+import calculateDynamicRate from "./util/calculateDynamicRate.js";
+
+const packageName = "fulfillment-method-shipping-dynamic-rate";
+const fulfillmentTypeName = "shipping";
+const fulfillmentMethodName = "dynamicRate";
+
+/**
+ * @summary Returns a list of fulfillment method quotes based on the items in a fulfillment group.
+ * @param {Object} context - Context
+ * @param {Object} commonOrder - details about the purchase a user wants to make.
+ * @param {Array} [previousQueryResults] - an array of shipping rates and
+ * info about failed calls to the APIs of some shipping methods providers
+ * e.g Shippo.
+ * @returns {Array} - an array that contains two arrays: the first array will
+ * be an updated list of shipping rates, and the second will contain info for
+ * retrying this specific package if any errors occurred while retrieving the
+ * shipping rates.
+ * @private
+ */
+export default async function getFulfillmentMethodsWithQuotesShippingDynamicRate(context, commonOrder, previousQueryResults = []) {
+  const { collections: { Fulfillment } } = context;
+  const [rates = [], retrialTargets = []] = previousQueryResults;
+  const currentMethodInfo = { packageName };
+
+  if (retrialTargets.length > 0) {
+    const isNotAmongFailedRequests = retrialTargets.every((target) => target.packageName !== packageName);
+    if (isNotAmongFailedRequests) {
+      return previousQueryResults;
+    }
+  }
+
+  const shippingRateDocs = await Fulfillment.find({
+    "shopId": commonOrder.shopId,
+    "fulfillmentType": fulfillmentTypeName,
+    "provider.enabled": true
+  }).toArray();
+  if (!shippingRateDocs || !shippingRateDocs.length) {
+    return [rates, retrialTargets];
+  }
+  const initialNumOfRates = rates.length;
+
+  shippingRateDocs.map(async (doc) => {
+    const carrier = doc.provider.label;
+    const currentPluginMethods = doc.methods.filter((method) => ((method.fulfillmentMethod === (fulfillmentMethodName)) && (method.enabled)));
```

Comment:
Why is the reasoning behind persisting the fulfillment method for a dynamic method? If the rates for are method are dynamically calculated on every `fulfillmentMethodQuotes` by calling an external API I don't see much point in having a method with predefined rates here. It can be dynamically constructed on every get quotes request. The only reasoning I can think of is i we allow the method to be enabled or not. But in that case we should check that before pushing the newly calculated rate. The same goes of all other method plugins.

Reply:
Not sure if I completely understood your point. The idea is that we are showing a placeholder function here which could be replaced by an actual API invoking function to collect the dynamic rate. The placeholder function is a dummy and currently returning some hard-coded values. It is expected to collect the actual dynamic quote/rate each time. Now the call to `fulfillmentMethodQuotes` will be placed to all ff-methods defined under the specific fulfillment-type (like all shipping or all pickup) and this is collected and presented to the user to select one of them as the  preferred method via the `selectFulfillmentOptionForGroup` call.

We can have a quick call if I got your question wrong.